flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/AbstractUserClassPathJobGraphRetriever.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.entrypoint.component;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.FileVisitOption;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ *  Abstract class for the JobGraphRetriever, which wants to get classpath user's code depends on.
+ */
+public abstract class AbstractUserClassPathJobGraphRetriever implements JobGraphRetriever {
+
+	protected static final Logger LOG = LoggerFactory.getLogger(AbstractUserClassPathJobGraphRetriever.class);
+
+	/** The directory contains all the jars, which user code depends on. */
+	@Nullable
+	private final String jobDir;
+
+	private List<URL> userClassPaths;
+
+	public AbstractUserClassPathJobGraphRetriever(String jobDir) {
+		this.jobDir = jobDir;
+	}
+
+	protected List<URL> getUserClassPaths() throws IOException {
+		if (userClassPaths == null) {
+			userClassPaths = getRelativeJarsURLFromDir(jobDir);
+		}
+		return userClassPaths;
+	}
```

Comment:
Such a method is usually an indicator that this class does too many things. Why don't we initialize this class with `userClassPaths` instead of calculating it within this class?

Reply:
At first, I initialize this class with `userClasspaths` but it needs some i/o operation which might  throw some `IOException`. This leads to some interface signature changes.  For example adding a `throw IOException` to `StandaloneJobClusterEntryPoint::createDispatcherResourceManagerComponentFactory`

However, I think we can initialize the class with `userClasspaths` and wrapper the `IOException` to some `RuntimeExcpetion` for example `IllegalArgumentException`.

Reply:
Maybe it would be easier to let `createDispatcherResourceManagerComponentFactory` throw an `IOException`. Given that we need to list some directories, this is fair enough.